src/bin/options.ml

```diff
@@ -268,16 +269,183 @@ let report_style =
+  ) (Dolmen_model.Ext.list ())
+
+let builtin_extensions =
+  merge_extensions
+    (List.rev_append builtin_typing_extensions builtin_model_extensions)
+
+let all_extensions, invalid_extensions =
+  let add_plugin ok err plugin = function
+    | None -> ok, plugin :: err
+    | Some (extension_name, k) ->
+      let typing_plugin =
+        match k with
+        | None | Some `Typing -> Dune_plugin plugin
+        | Some _ -> Unavailable
+      and model_plugin =
+        match k with
+        | None | Some `Model -> Dune_plugin plugin
+        | Some _ -> Unavailable
+      in
+      { extension_name ; typing_plugin ; model_plugin } :: ok, err
+  in
+  let ok, err =
+    List.fold_left (fun (ok, err) e ->
+      add_plugin ok err e @@ parse_ext_opt e
+    ) (builtin_extensions, []) (Dolmen.Sites.Plugins.Plugins.list ())
+  in
+  merge_extensions ok, List.fast_sort String.compare err
+
+let pp_extension ppf e =
+  let has_typing = plugin_is_available e.typing_plugin in
+  let has_model = plugin_is_available e.model_plugin in
+  let variants =
+    (if has_typing then ["typing"] else []) @
+    (if has_model then ["model"] else [])
+  in
+  Fmt.pf ppf "%s@ %a" e.extension_name
+    Fmt.(parens @@ list ~sep:comma string) variants
+
+let find_ext name =
+  try Ok (List.find (fun e -> e.extension_name = name) all_extensions)
+  with Not_found ->
+    Fmt.error_msg
+      "@[<v>Could not find extension '%s'@ \
+      Available extensions:@;<1 2>@[<v>%a@]@]"
+      name
+      Fmt.(list (box pp_extension)) all_extensions
+
+let load_typing_extension ext =
+    match ext.typing_plugin with
+    | Unavailable ->
+      Fmt.error_msg "No typing extension '%s'" ext.extension_name
+    | Builtin e ->  Ok e
+    | Dune_plugin plugin ->
+      Dolmen.Sites.Plugins.Plugins.load plugin;
+      try
+        Ok (List.find
+          (fun e -> Dolmen_loop.Typer.Ext.name e = ext.extension_name)
+          (Dolmen_loop.Typer.Ext.list ())
+        )
+      with Not_found ->
+        Fmt.error_msg
+          "Plugin '%s' did not register a typing extension for '%s'"
+          plugin ext.extension_name
+
+let load_model_extension ext =
+    match ext.model_plugin with
+    | Unavailable ->
+      Fmt.error_msg "No model extension '%s'" ext.extension_name
+    | Builtin e ->  Ok e
+    | Dune_plugin plugin ->
+      Dolmen.Sites.Plugins.Plugins.load plugin;
+      try
+        Ok (List.find
+          (fun e -> Dolmen_model.Ext.name e = ext.extension_name)
+          (Dolmen_model.Ext.list ())
+        )
+      with Not_found ->
+        Fmt.error_msg
+          "Plugin '%s' did not register a model extension for '%s'"
+          plugin ext.extension_name
+
+let extension =
+  let parse s =
+    match parse_ext_opt s with
+    | None -> Fmt.error_msg "Invalid extension name '%s'" s
+    | Some (name, kind) ->
+      Result.map (fun ext -> (ext, kind)) (find_ext name)
+  in
+  let print ppf _p = Fmt.pf ppf "" in
+  Arg.conv (parse, print)
```

Comment:
I think this is enough code to warrant it being mostly in a new file (`extensions.ml` ?).

Reply:
Fair enough, will do.